Sign Up

MathiasLantean opened the pull request:
This end-point allows users to do CRUD operations (old API, before talking to Kalil).
Sign up, sign in, logout and email verification .

Commit:
Add custom User model and basic API

Specify URL organization.

Commit:
Rename file

In order to match the convention

Commit:
Add password to the User serializer method 

The password is only an editable field, there is no way to get the attribute through the API.
Added missing PK attribute.

Commit:
Add basename to URLs

Commit:
Add test

The test structure was modified.

Commit:
Modify app name

As another third-party package (allauth) has the app name 'account' taken. The app created with this name was refactored.

Commit:
Set dj-rest-auth

As dj-rest-auth needs rest_framework and allauth, some configurations are from these packages.

Commit:
Add test

As allauth will be used for authentication, some pieces of code didn't make sense, they were removed.

kaozdl commented on profile/models.py:
```diff
@@ -0,0 +1,34 @@
+from __future__ import unicode_literals
```
this is not needed, in python3 all strings are unicode by default

MathiasLantean replied:
You are right.

kaozdl commented on profile/serializers.py:
```diff
@@ -0,0 +1,144 @@
+UserModel = get_user_model()
+
+
+class UserSerializer(serializers.HyperlinkedModelSerializer):
```
Why are you using HyperlinkedModelSerializer?

This generates a url field for the related models instead of a Foreign Key, but you do not seem to have any related model.

MathiasLantean replied:
I totally agree. At the beginning, I was using a one to one relation between Profile model and Django User model. After changing this for a custom User model instead of using the Django User I forgot to update the serializer. It could use just `serializers.ModelSerializer`

kaozdl reviewed (commented):
So far looks good.
Left some comments

asermax commented on profile/serializers.py:
```diff
@@ -0,0 +1,144 @@
+        extra_kwargs = {'password': {'write_only': True, 'min_length': 4}}
+
+
+class UserDetailsSerializer(serializers.ModelSerializer):
```
Do you need a separate serializer to handle the details? Since the `password` field is write only on the serializer above, you should be able to use the same one both for deserialization and serialization.

MathiasLantean replied:
Well, extra_kwargs are not shown in the details serializer. This serializer accepts `password` as a parameter if the method is POST, UPDATE or PATCH but is not added to the serialized data in the GET method. Any way this API will be removed soon as dj-rest-auth was installed.

asermax commented on profile/views.py:
```diff
@@ -0,0 +1,12 @@
+User = get_user_model()
+
+
+class UserViewSet(viewsets.ModelViewSet):
```
Does this API need any kind of authentication? What about authorization, is anyone able to create/modify/delete any users?

MathiasLantean replied:
Good point, there is no authentication verification. This API was just a CRUD, it was because of that dj-rest-auth was installed. This API will be deprecated soon.

guilleijo commented on profile/serializers.py:
```diff
@@ -0,0 +1,144 @@
+        if username and password:
+            user = self.authenticate(username=username, password=password)
+        else:
+            msg = _('Must include "email" and "password".')
```
Shouldn't be "username" instead of "email" here?

guilleijo replied:
Shouldn't be "username" instead of "email" here?

guilleijo commented on profile/serializers.py:
```diff
@@ -0,0 +1,144 @@
+            user = self.authenticate(email=email, password=password)
+        else:
+            msg = _('Must include either or "email" and "password".')
+            raise exceptions.ValidationError(msg)
```
Could it be that an `elif username and password:` is missing here?

guilleijo commented on profile/serializers.py:
```diff
@@ -0,0 +1,144 @@
+    email = serializers.EmailField(required=allauth_settings.EMAIL_REQUIRED,
+                                   max_length=get_username_max_length(),
+                                   min_length=allauth_settings.USERNAME_MIN_LENGTH,
+                                   )
```
We are thinking in creating a style guide and adopt the indentation suggested by django in [coding style](https://docs.djangoproject.com/en/3.0/internals/contributing/writing-code/coding-style/) and use hanging indentation rather than vertical alignment:
```
email = serializers.EmailField(
    required=allauth_settings.EMAIL_REQUIRED,
    max_length=get_username_max_length(),
    min_length=allauth_settings.USERNAME_MIN_LENGTH,
)
```
What do you think about that?

guilleijo commented on profile/serializers.py:
```diff
@@ -0,0 +1,144 @@
+        return data
+
+    def custom_signup(self, request, user):
+        pass
```
Is this necessary?

kaozdl replied:
this method is meant to add extra functionality.
if it's not used i think it can be deleted

guilleijo commented on profile/tests/test_auth.py:
```diff
@@ -0,0 +1,83 @@
+from ..models import User
+
+
+class UserAuthTests(APITestCase):
```
Seeing that you are using some test values like "test@test.com" and "test123455" multiple times, maybe you could define them in a `setUpTestData` class method.

guilleijo commented on profile/tests/tests_managers.py:
```diff
@@ -0,0 +1,42 @@
+        self.assertRaises(ValueError, self.manager.create_user, email="", password="test1234", gender=2)
+
+    def test_can_create_user_without_gender(self):
+        # user = self.manager.create_user("", "password1234", gender=User.Gender.GENDER_MALE)
```
Should these comments be deleted?

guilleijo reviewed (commented):
Left some comments 👍

Commit:
Code review fixes

Remove unused imports.
Update settings so email verification is optional.
Update profile serializers.
Improve tests.
